### ISSUE TITLE
Distributed SQL Support

### ISSUE THREAD

**Issue by sjezewski** [pachyderm/pachyderm#3066]
**What is the ask?**

As a user, I need to run large / complicated / unwieldly SQL queries on my data in Pachyderm

---

Broadly, the deliverable will look like the following. For the first pass, we'll assume postgres. Later on we'll add support for other adapters/formats.

1) Basic Ingress Support

- Update `put-file --split` to take `sql` as a delimiter option
- This will take as an input a `pgdump` file, and break it up into smaller valid `pgdump` files (so that include the schema/table setup at the start) which contain a subset of the rows (via insert statements) of the original file (the count depending on the other `--split` flags specified)

2) SQL engine sidecar

- A lot of complexity was added to an internal prototype to try and handle postgres management and user code all in a single container
- If you specify that the pipeline is a `sql` pipeline, it'll run this sidecar. Probably this will be an `adapter` field on the `transform` object. For now the only valid value will be `postgres`

3) First class worker support

- Once a pipeline is marked as having an `"adapter":"postgres"` the workers will ingest the data directly into the DB sidecar (not as files under `/pfs`)
- We'll also need to supply fields on the transform so that a user can specify
    - any SQL statements that they need to run (either a statement to run or a path to a file containing a stored proc)
    - the select statement for grabbing the output
- The output will be emitted as a CSV w headers
- We'll also need to handle graceful setup/cleanup (esp in the event of user code failures) of the DB since it affects the long running postgres sidecar

# Nice to Haves

Out of scope / future iterations

- Ingress Pipeline
    - Cron ... periodically polls external database for ingress
    - Keeps track of state ... either w limit/offset/from/cursor
    - Configurable via secret (for DB URL/port/password/schema/table)
    - Emits data in the format that is expected for the DB worker
- Parallelism
    - Since we have a long running postgres instance alongside the worker, we can parallelize datum processing ...
    - If we make sure the DB uses namespaces
- Adv config
    - specify resource req/limits for the db sidecar separately (maybe this is once we have first class k8s manifests for pipelines)
- More adapters
    - MySQL / cockroach / etc

**Comment by gabrielgrant** [comment]
Another config request: allowing export as the same PG dump format that we take for input, so that multiple postgres steps can be chained

**Comment by jdoliner** [comment]
> Another config request: allowing export as the same PG dump format that we take for input, so that multiple postgres steps can be chained

This should be the default, and probably only, option for SQL pipelines.

**Comment by jdoliner** [comment]
Good point, CSV and SQL dump are the 2 main formats we want to support. CSV should be usable for chaining SQL pipelines as well.

**Comment by gabrielgrant** [comment]
The one kinda tricky issue is that, although widely used, CSV is a pretty crappy format, with a ton of different subtly-different implementations, none of which have nearly the level of typing support that a SQL DB does (technically every value is a string). So we may also want to support an object-per-row stream of JSON objects:

```
{"id":1,"column_1":43,"column_2":"asfasfasfd"}
{"id":2,"column_1":43,"column_2":"asfasfasfd"}
...
```

(this is what is output by pg's `ROW_TO_JSON`, and already supported with our `put-file --split json`)

**Comment by gabrielgrant** [comment]
More generally, this seems like a pretty powerful pattern (would be super useful for the spark stuff I've been working on). Could we generalize this into a "pipeline-data adapter API", so the transform could take:

- a sidecar image
- an input-data PFS-to-adapted-pipeline image
- an output-data adapted-pipeline-to-PFS image

so that we have something to point people at, rather than getting ourselves into the black hole of having to build and bake in every possible adapter ourselves?

(presumably we would want to build some of these, like postgres, ourselves with more official support/sugar, but having this seems like a good extension point around which to hopefully increase community contributions)

**Comment by gabrielgrant** [comment]
Another consideration for the pg-specific case is that, for a multi-stage SQL pipeline, dumping and loading at every stage can be incredibly time consuming (especially if rebuilding secondary indices). This quite possibly outside the scope of what we should do for a first pass at this feature, but I started documenting some possible patterns for how integration with postgres could here (warning: this is pretty old; demo code is incomplete/almost-certainly broken): https://github.com/gabrielgrant/pach-postgres-demo

**Comment by jdoliner** [comment]
> The one kinda tricky issue is that, although widely used, CSV is a pretty crappy format, with a ton of different subtly-different implementations

I don't disagree with this, although I don't have quite the level of CSV hate that @gabrielgrant does, but I think it's a common enough format that we're going to need to support it sooner or later. JSON can also be made to work with our SQL support. Parquet is probably going to be necessary eventually as well... but let's not expand to scope of this issue any further.

 > More generally, this seems like a pretty powerful pattern (would be super useful for the spark stuff I've been working on). Could we generalize this into a "pipeline-data adapter API"

Definitely, that's the direction this is all going, we'll be keeping an eye on how to generalize this but I don't want to creep the scope of this too much to support everything in one PR. But we'll make sure not to do anything that will make it harder to support in the future. Re: the different containers @gabrielgrant describe. Those will likely going in our Pipeline definition as `Extract` and `Load` fields to match our existing `Transform` field.

> Another consideration for the pg-specific case is that, for a multi-stage SQL pipeline, dumping and loading at every stage can be incredibly time consuming

This would have some performance benefits, but I don't think this is as clear a performance win as you think it is. A couple of things that mitigate it:

- we parallelize that loading process, and it will often be a relatively small part of the total runtime compared to the processing
- this benefits from deduplication, so you generally won't need to load the same data twice
- downstream pipelines being able to consume data directly out of upstream pipeline's database, while more efficient, creates a coupling between 2 pipelines that didn't exist. Allowing load from one pipeline to bring down the other.
- the amount of data being loaded into the database per datum may be very small, as small as a single row, in that case optimizing how we put it will have very little effect and the other costs associated with a datum are going to dominate.

Just generally I think we should punt on performance related goals for this project save for not implementing things that we know will make it impossible to get reasonable performance out of this system going forward. Caching things rather than reloading them is a great optimization to make later on, because regardless of what we do, we're going to need all of the code that loads the data in to the DBs and extracts it. Since these DBs that we're keeping up as caches need to come from somewhere. These types of optimizations will make sense at some point, probably when we're getting user pressure to make a workload more efficient.

> Higher level question: in cases where people really want to distribute big SQL queries, wondering if we should maybe consider just recommending/integrating something like Citus

I think this plays pretty nice with our side car model, right now we're thinking of the side car as a single container, but it could also be a cluster of containers and the API wouldn't have to change much if at all to support it. I will say though that this isn't quite letting each system play to their strengths, Citus and Cockroach are both designed as distributed SQL for _transactions_, we're doing distributed SQL for _analysis_, the major difference is that we know a priori exactly how much data we have and can chunk it up however we want ahead of time, Citus and Cockroach are designed to receive a constant stream of inserts and distribute those live with low latency while keeping storage and compute usage fairly even between nodes. That's a much harder problem than the one we're looking for a solution to. Presto and Hive are probably the best examples of what we'd like an outside system to look like... unfortunately they're both written in Java and bring over some of the crappiness of the Hadoop ecosystem, but it may be fine integrating them since they'll be off in their own containers, doing their own thing. If there was a go implementation of Presto that would really be the best case scenario for us, but I don't think it really exists.
